solidigm: Added parsing for telemetry customer screenable data

lgdacunh opened the pull request.

lgdacunh commented:
@igaw, I am the same developer that contributed the other solidigm plug-in commits. For this pull request I am using a new GitHub account, created according to Solidigm polices, to join the Solidigm organization in GitHub. Can you please approve my new user to run workflows?

igaw commented on plugins/solidigm/solidigm-telemetry/telemetry-log.h:
```diff
@@ -11,10 +11,23 @@
+    typedef struct { int CONCATENATE_VAR_NAMES(static_assertion_failed__,msg) : !!(cond); } \
+        EVALUATE_CONCATENATE_VAR_NAMES(static_assertion_failed_,__COUNTER__)
+
+#define MEMBER_SIZE(type, member) sizeof(((type *)0)->member)
```
I'd like to avoid this macro madness. What about exploiding the fact we are using C99 which allows to use `_Static_assert`?

```c
#include <assert.h>

#if !defined __cplusplus
#define static_assert _Static_assert
#endif

static_assert(sizeof(unsigned int) == 4, "no scope");

int main(int argc, char *argv[])
{
	static_assert(sizeof(unsigned int) == 4, "inside main");
	return 0;
}
```

And then write this just as 

```c
static_assert(sizeof(struct reason_indentifier_1_0) == member_size(...,...), "size miss match");
```

and the `member_size` macro should definitely be in utils.h

lgdacunh commented on plugins/solidigm/solidigm-telemetry.c:
```diff
@@ -19,12 +19,40 @@
+{
+	FILE * fd = fopen(file_name, "rb");
+	if (!fd)
+	{
```
Nice suggestion! I wasn't aware of checkpatch.pl

lgdacunh commented on plugins/solidigm/solidigm-telemetry.c:
```diff
@@ -33,63 +61,110 @@ int solidigm_get_telemetry_log(int argc, char **argv, struct command *cmd, struc
 
 	if (cfg.host_gen > 1) {
-		fprintf(stderr, "Invalid host-generate value '%d'\n", cfg.host_gen);
+		SOLIDIGM_LOG_WARNING("Invalid host-generate value '%d'\n", cfg.host_gen);
```
I will work on cleaning this up. I would like to have an abstraction that allows me to redirect the warnings without messing up with stderr.

igaw commented:
I think you want to use the helpers from

https://github.com/linux-nvme/nvme-cli/blob/3edbfd0dbe049dedf7259ed97bf40a679892a024/util/json.h#L19

And while at it, please shorten those overlong lines.

lgdacunh commented:
> I think you want to use the helpers from
> 
> https://github.com/linux-nvme/nvme-cli/blob/3edbfd0dbe049dedf7259ed97bf40a679892a024/util/json.h#L19
> 
> And while at it, please shorten those overlong lines.

@igaw, I am missing helper code to enable json_object_array_put_idx() of uin64 values. I will do something with a local #define, as this doesn't seen needed in other parts of the code.

lgdacunh commented:
> > I think you want to use the helpers from
> > https://github.com/linux-nvme/nvme-cli/blob/3edbfd0dbe049dedf7259ed97bf40a679892a024/util/json.h#L19
> > 
> > And while at it, please shorten those overlong lines.
> 
> @igaw, I am missing helper code to enable json_object_array_put_idx() of uin64 values. I will do something with a local #define, as this doesn't seen needed in other parts of the code.

@igaw, I changed my mind here, and changed the header file. The least places we branch code based on CONFIG_JSONC_14  the better.
I already shortened the very long lines.

igaw commented:
I've merged this PR directly because I can't update this PR. 

Anyway, I've spend a bit of time just to fix the most offensive coding style issues. This is the second time I've done this and it's the last time. Next time you have to fix it up yourself until it's an acceptable state. I know everyone loves her/his own coding style. Though this is shared code, this means we need to stay consistent to make it easier to consume later.

For example while reformatting the parser code I noticed immediately that the error code is leaking memory all over the place. Before the reformatting I didn't saw it.

Just to iterate it again: stick to the coding guidelines this makes reading the code for trained eyes easy.
